### PR TITLE
feat: kv query callbacks

### PR DESCRIPTION
**Pull request by zavgorodnii**
[closes https://p2pvalidator.atlassian.net/browse/LSC-110]

this PR introduces optional callbacks for KV queries. the `QueryResult` message now has the boolean `allow_kv_callbacks` field; if set to true, neutron `sudo()`'es a message with the query id to the contract that owns the query. Note: the message only contains the query ID, contracts have to load the data manually after receiving the notification.

this PR depends on:

1.  https://github.com/neutron-org/neutron-contracts/pull/9
2. https://github.com/neutron-org/cosmos-query-relayer/pull/18

**HOW TO TEST THIS CODE?**

You can find the instructions here: https://github.com/neutron-org/neutron-contracts/blob/feat/kv-query-callbacks/contracts/neutron_interchain_queries/README.md#testing-the-sudomessagekvqueryresult-handler

### CODE REVIEW

File: internal/sudo/sudo.go [hunk 1]
```diff
@@ -215,12 +237,50 @@ func (s *SudoHandler) SudoTxQueryResult(
...
+	if err != nil {
+		s.Logger(ctx).Debug("SudoTxQueryResult: failed to Sudo",
+			"error", err, "contract_address", contractAddress)
+		return nil, err
+	}
+
+	return resp, nil
+}
+
+func (s *SudoHandler) SudoKVQueryResult(
```
Comment (mmikev) [inline internal/sudo/sudo.go:257]: Descriptive comment would be nice

File: x/interchainqueries/keeper/msg_server.go [hunk 2]
```diff
@@ -72,6 +72,13 @@ func (k msgServer) SubmitQueryResult(goCtx context.Context, msg *types.MsgSubmit
 		return nil, sdkerrors.Wrapf(err, "failed to get query by id: %v", err)
 	}
 
+	queryOwner, err := sdk.AccAddressFromBech32(query.Owner)
+	if err != nil {
+		ctx.Logger().Error("SubmitQueryResult: failed to decode AccAddressFromBech32",
+			"error", err, "query", query, "message", msg)
+		return nil, sdkerrors.Wrapf(types.ErrInternal, "failed to decode owner contract address: %v", err)
```
Comment (mmikev) [inline x/interchainqueries/keeper/msg_server.go]: `sdkerrors.Wrapf(err, ...)`?

File: internal/sudo/sudo.go [hunk 3]
```diff
@@ -215,12 +237,50 @@ func (s *SudoHandler) SudoTxQueryResult(
...
+	contractAddress sdk.AccAddress,
+	queryID uint64,
+) ([]byte, error) {
+	s.Logger(ctx).Info("SudoKVQueryResult", "contractAddress", contractAddress)
+
+	// TODO: basically just for unit tests right now. But i think we will have the same logic in the production
+	if !s.wasmKeeper.HasContractInfo(ctx, contractAddress) {
+		s.Logger(ctx).Debug("contract was not found", "contractAddress", contractAddress)
+		return nil, nil
+	}
```
Comment (foxpy) [inline internal/sudo/sudo.go:268]: Seems to be logical to have this logic here. I suggest deleting this comment.

Comment (zavgorodnii) [reply internal/sudo/sudo.go:268]: We will deal with all comments of this type in a separate task.

**Review by foxpy — DISMISSED** [review]

**Review by swelf19 — APPROVED** [review]

**Review by pr0n00gler — APPROVED** [review]

**Review by foxpy — APPROVED** [review]

**Pull request merged**